### PR TITLE
MNT: Reconcile docstrings and function signatures

### PR DESCRIPTION
**Pull request by rmcgibbo**
Based on #4575, I wrote a [small tool](https://github.com/rmcgibbo/doccheck) to check for discrepancies between function signatures and docstrings. This PR cleans up some of them.

Extra notes: 
1. Docstring for `Akima1DInterpolator` mentions `axis` as argument, but not in signature.
2. Docstring for `NearestNDInterpolator` is quite different than signature: `(x,y)` vs. `points`
3. Not sure what's going on  with `spleval`
4. `scipy.interpolate.PiecewisePolynomial.extend`'s docstring's `direction` argument doesn't appear in signature.
5. `brute_force` argument in `scipy.ndimage.morphology.binary_erosion` and `scipy.ndimage.morphology.binary_dilation` undocumented.

### CODE REVIEW

**Comment by ewmoore** [comment]
Do you think it's worth adding your tool somewhere? I think we have a tools
directory. Alternative it could possibly go in to numpy doc somewhere.

**Comment by rmcgibbo** [comment]
Numpydoc is probably an appropriate place for it, if the maintainers are
interested, now that it's a separate project from numpy core.

The tool still needs to be cleaned up a little bit (it only runs on py3k,
not documented, etc).

**Comment by rmcgibbo** [comment]
There's kind of a lot here, and it touches quite a few files. If the maintainers would prefer that I split this up into separate PRs or something, I'd be happy to do that.

File: scipy/linalg/decomp.py [hunk 1]
```diff
@@ -778,7 +778,7 @@ def eigvals_banded(a_band, lower=False, overwrite_a_band=False,
         The bands of the M by M matrix a.
     lower : boolean
         Is the matrix in the lower form. (Default is upper form)
-    overwrite_a_band:
+    overwrite_a_band : boolean
```
Comment (rgommers) [inline scipy/linalg/decomp.py]: ``bool`` not ``boolean``. 

The rule is actual types here (so ``int, bool, ndarray``) and clear English in the text (so ``integer, boolean, array``).

Comment (rgommers) [inline scipy/linalg/decomp.py]: this comment applies to a few more usages of ``boolean`` in this PR.

File: scipy/signal/ltisys.py [hunk 2]
```diff
@@ -628,7 +628,7 @@ def lsim(system, U, T, X0=None, interp=1):
     T : array_like
         The time steps at which the input is defined and at which the
         output is desired.
-    X0 :
+    X0 : array-like
```
Comment (rgommers) [inline scipy/signal/ltisys.py]: ``array_like``, with underscore

File: scipy/sparse/compressed.py [hunk 3]
```diff
@@ -127,12 +127,10 @@ def check_format(self, full_check=True):
...
 
         Parameters
-        ==========
-
-            - full_check : {bool}
-                - True  - rigorous check, O(N) operations : default
-                - False - basic check, O(1) operations
-
+        ----------
+        full_check : bool
```
Comment (rgommers) [inline scipy/sparse/compressed.py]: add ``, optional``

File: scipy/spatial/distance.py [hunk 4]
```diff
@@ -1611,7 +1611,7 @@ def is_valid_y(y, warning=False, throw=False, name=None):
         condensed distance matrix. The warning message explains why
         the distance matrix is not valid.  `name` is used when
         referencing the offending variable.
-    throws : throw, optional
+    throw : throw, optional
```
Comment (rgommers) [inline scipy/spatial/distance.py]: type identifier should be ``bool``

**Comment by rgommers** [comment]
No need to split this PR, it looks good to me as is. Can be merged after the few comments I made are adressed.

**Comment by rgommers** [comment]
@rmcgibbo your "extra notes " in the PR description are a request for someone to check/fix these right? I can have a look at them.

**Comment by rmcgibbo** [comment]
> @rmcgibbo your "extra notes " in the PR description are a request for someone to check/fix these right? I can have a look at them.

Yeah. I wasn't really familiar with those functions, but I thought someone who is might want to consider them.

File: scipy/linalg/decomp.py (hunk 1 above)
Comment (rmcgibbo) [inline scipy/linalg/decomp.py]: I fixed up a lot more of these https://github.com/rmcgibbo/scipy/commit/6e71cf0baba66d3df0e378ad41b1311b96b33c64

**Comment by ev-br** [comment]
A couple of "extra notes" I can comment on:

1. Akima1DInterpolator, axis argument: not implemented as of master. Changing it here would probably create a merge conflict with gh-4486 (which implements it), but  that conflict would be easy to deal with.

3. `spleval`/`splmake` are well hidden, slated for deprecation in https://github.com/scipy/scipy/issues/1408. Probably best be left hidden.

**Comment by rmcgibbo** [comment]
There's still some inconsistency, especially in `linalg`, on the way that optional bool parameters (like `check_finite`, `overwrite_X`) are written. Many are missing the optional. I'm going to update the tooling to check for this and make one more pass through.

**Comment by rmcgibbo** [comment]
There were more of these than I expected...

**Comment by rgommers** [comment]
Merged. Thanks @rmcgibbo, this is great!

**Comment by rgommers** [comment]
Regarding the ``brute_force`` argument (5), that selects a completely different algorithm defined in ``ni_morphology.c`` with zero documentation. I don't understand what it does or why by reading the code. I'll open an issue for that one.

**Comment by rgommers** [comment]
``NearestNDInterpolator`` is odd - the ``__init__`` signature has been made inconsistent with the class it inherits from on purpose. It looks like the docstring needs adapting but I'm not 100% sure. @pv?

**Comment by rgommers** [comment]
``direction`` looks like it needs to be removed from ``PiecewisePolynomial.extend``. It's also not present in ``append``, which uses ``self.direction`` instead.

**Comment by rgommers** [comment]
Addresses the ``NearestNDInterpolator`` and ``PiecewisePolynomial.extend`` issues in gh-4609.

**Comment by rgommers** [comment]
That takes care of all the loose ends here I think.

**Comment by JDWarner** [comment]
@rgommers I'm taking a look at the binary morphology routines in `scipy.ndimage` and it appears `origin` and `border_value` remain out of order in several docstrings.

**Comment by rgommers** [comment]
@JDWarner indeed. I think that due to the missing ``brute_force`` parameter, @rmcgibbo's tool didn't trigger on the order. I'll add your comment to gh-4586, fits better there than here.

**Pull request merged**